text/0000-typed-context-injection.md

```diff
@@ -0,0 +1,833 @@
+
+```rust
+fn func_1(..., new_system: &mut NewSystem) {  // Updated!
+    ...
+    func_2(..., new_system);  // Updated!
+}
+
+fn func_2(..., new_system: &mut NewSystem) { // Updated!
+    ...
+    func_3(..., new_system);  // Updated!
+}
+
+fn func_3(..., new_system: &mut NewSystem) { // Updated!
+    // We finally have access to `new_system`
+}
+```
+
+Although one may be tempted to bundle up this context into convenient tuples to reduce the amount of explicit context passing, doing so will only work if at most one descendant of the call stack relies upon that type. If several descendants rely on a given contextual type, you're out of luck!
+
+```rust
+type Descendant1Cx<'a> = (&'a mut SharedSystem, &'a mut Descendant1System);
+
+fn descendant_1(cx: Descendant1Cx<'_>) {
+    ...
+}
+
+type Descendant2Cx<'a> = (&'a mut SharedSystem, &'a mut Descendant2System);
+
+fn descendant_2(cx: Descendant2Cx<'_>) {
+    ...
+}
+
+fn callee(descendant_1_cx: Descendant1Cx<'_>, descendant_2_cx: Descendant2Cx<'_>) {
+    descendant_1(descendant_1_cx);
+    descendant_2(descendant_2_cx);
+}
+
+fn caller(
+    shared_system: &mut SharedSystem,
+    descendant_1_system: &mut Descendant1System,
+    descendant_2_system: &mut Descendant2System,
+) {
+    callee(
+        (shared_system, descendant_1_system),
+        (shared_system, descendant_2_system),  // Whoops! We borrowed shared_system twice!
+    );
+}
+```
+
+Finally, although good software practices encourage programmers to keep their abstractions minimal and properly-decoupled, doing so only causes grief when trying to implement operations acting upon several of these well-split-up components:
+
+```rust
+pub struct AudioBuffer { ... }
+
+impl AudioBuffer {
+    pub fn play_samples(&mut self, data: &[f32], volume: f32) { ... }
+}
+
+pub struct VirtualFileSystem { ... }
+
+impl VirtualFileSystem {
+    pub fn load_file(&mut self, path: &str) -> &[u8] { ... }
+}
+
+pub struct AudioCache { ... }
+
+impl AudioCache {
+    pub fn resolve_from_cache(&mut self, path: &str, loader: impl FnOnce() -> &[u8]) -> &[f32] { ... }
+}
+
+pub struct SoundVolumeOverrides { ... }
+
+impl SoundVolumeOverrides {
+    pub fn get_volume_override(path: &str) -> f32 { ... }
+}
+
+// Ideally, we could just write:
+cx.play_sound_at_path("cbat.ogg");
+
+// But, unfortunately, we must write:
+audio_buffer.play_sound_at_path(&mut virtual_fs, &mut audio_cache, &mut sound_volume_overrides, "cbat.ogg");
+```
+
+Yuck!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Enter typed context injection. 
+
+Context injection is accomplished with two new standard library items: `Cx` and `AnyCx`. `Cx` is a variadic type parameterized by the references comprising a function's context. For example, to write a function taking `&mut System1` and `&System2`, you can just write:
+
+```rust
+fn consumer(cx: Cx<&mut System1, &System2>) { ... }
```

Comment:
Users can still use `Cx` to build their own custom context tuples. For example, an ECS could define its own `AccessCx` type alias which resolves to a `Cx` object with appropriately typed access tokens. I don't want to allow users to define their own `Cx` objects because doing so would prevent useful interoperability between different context types.

With the ECS example, handling `AccessCx` and regular `Cx` objects the same way could allow injecting both regular object references and access tokens in the same context like so:

```rust
fn my_consumer(cx: Cx<AccessCx<&mut Component1, &Component2>, &MyDependency>) {}
```
...which would expand to:

```rust
fn my_consumer(cx: Cx<Cx<&mut AccessToken<Component1>, &AccessToken<Component2>>, &MyDependency>) {}
```

Reply:
Why do you keep talking of "type aliases"? Type aliases are supposed to be indistinguishable from their values. Meanwhile, here you seem to be mixing up type equality and subtyping.

Reply:
Inability to interop between different context types is the whole point why I am mentioning this. It's just like you wouldn't want to transparently pass the resource typemap from `rocket` to the typemap in `serenity` just because "they are both contexts".

Reply:
> Inability to interop between different context types is the whole point why I am mentioning this. It's just like you wouldn't want to transparently pass the resource typemap from `rocket` to the typemap in `serenity` just because "they are both contexts".

I think part of your misunderstanding of this proposal comes from the fact that `Cx` is not supposed to be enumerable. `Cx` refers to an exact type—the tuple of component references you wish to accept. If you don't want a component, you won't get it.

Hence, if Rocket requests a context of components `&RocketResource<Thing1>` and `&RocketResource<Thing2>`, and the user has a context `Cx<&RocketResource<Thing1>, &RocketUserResource<Thing2>, &RocketResource<Thing3>, &SerenityResource<Thing4>>`, it will just receive `&RocketResource<Thing1>` and `&RocketResource<Thing2>`—nothing more, nothing less. Hence, it is not dangerous to mix contexts.

> Why do you keep talking of "type aliases"? Type aliases are supposed to be indistinguishable from their values. Meanwhile, here you seem to be mixing up type equality and subtyping.

I'm not sure I understand what you mean by this.